### DISCUSSION TITLE
Find a commit by oid/sha via graphql (v4) api

### DISCUSSION THREAD

**Discussion by sorenlouv** [community/community#24528] [Apps, API and Webhooks]
<p>With the REST api it is possible to find a commit by the commit sha via this endpoint:</p>
<pre><code>https://api.github.com/search/commits?q=hash:${sha}%20repo:${repoOwner}/${repoName}
</code></pre>
<p>and this</p>
<pre><code>https://api.github.com/repos/${repoOwner}/${repoName}/git/commits/${sha}
</code></pre>
<p>How can I achieve the same via the graphql api? I tried using the search resouce but it seems like it only supports searching across issues, repositories and users - not commits:</p>
<pre><code>{
  search(query: "hash: e9e1b1f5021358f2c769da20579773688f7d5638" type: COMMIT) {
    ...
  }
}
</code></pre>
<p>Thanks in advance!</p>

**Comment by sorenlouv** [comment]
<p>I was able to <a href="https://developer.github.com/v4/guides/using-global-node-ids/" rel="nofollow noopener">find the global node id by using the REST api</a> and then getting the commit via the graphql api:</p>
<pre><code>{
  node(id: "MDY6Q29tbWl0NzgzMzE2ODowZjNmZWFlZGMyODBlY2FmMTE2NDE5MTkzNWRiMmYxNjdhMDc0NDgy") {
    id
    ... on Commit {
      message
    }
  }
}
</code></pre>
<p>This seems a bit counter-productive, since I’ve migrated to the graphql api to avoid over-fetching from the REST api. </p>
<p>Additional question: Is there any way to see which branch the commit belongs to? I can see that Github Web UI uses an internal API to do this:</p>
<p><a href="https://github.com/elastic/kibana/branch_commits/6a97f9a7ee797d302cb4f05120ff935bfe22eba0" rel="nofollow noopener">https://github.com/elastic/kibana/branch_commits/6a97f9a7ee797d302cb4f05120ff935bfe22eba0</a></p>

**Comment by wei** [comment]
<p>The v4 equilivent for</p>
<pre><code>https://api.github.com/repos/${repoOwner}/${repoName}/git/commits/${sha}
</code></pre>
<p>is </p>
<pre><code>{
  repository(owner: "&lt;repoOwner&gt;", name: "&lt;repoName&gt;") {
    object(oid: "&lt;sha&gt;") {
      ... on Commit {
        message
      }
    }
  }
}
</code></pre>

**Reply by babblebey** [reply]
Great stuff, this was everything.

**Comment by sorenlouv** [comment]
<p>This is great, thanks!</p>
<p>I was able to make it work for short sha’s as well:</p>
<pre><code>{
  repository(owner: "elastic", name: "kibana") {
    object(expression: "e9e1b1f5") {
      ... on Commit {
        message
      }
    }
  }
}
</code></pre>

**Comment by sorenlouv** [comment]
<p><a class="mention" href="/wei">@wei</a> Thanks again for your help.</p>
<p>I’m only interested in commits from master - do you know how to limit the results by branch (in addition to sha)?</p>
<p>Or alternatively: if it’s possible to see which branch the returned commit object belongs to? (that’ll allow me to discard non-master commits elsewhere in my application).</p>

**Comment by amyjbaer** [comment]
<p>I’m also looking to filter by commits to master. Were you able to figure this out <a class="mention" href="/sqren">@sqren</a> ? </p>

**Comment by sorenlouv** [comment]
<p>Not with v4 unfortunately. You can do it with v3: <a href="https://developer.github.com/v3/search/#considerations-for-commit-search" rel="nofollow noopener">https://developer.github.com/v3/search/#considerations-for-commit-search</a></p>

**Comment by jstacoder** [comment]
<p>Branch’s are just refs, like commits, and they are eaiser to get I think, because you just need to get it by it’s ref name</p>
<pre><code class="lang-auto">repository(name:\&lt;name\&gt; owner:\&lt;owner\&gt;){  
ref(qualifiedName:”refs/branchs/\&lt;branchName\&gt;”){}}  
</code></pre>
<p>And for master it’s even eaiser, because Repisitory has a defaultBranchRef field that usually points to master.</p>

**Comment by sorenlouv** [comment]
<p>Hey <a class="mention" href="/jstacoder">@jstacoder</a></p>
<p>Thanks for chiming in. I tried your suggestion like:</p>
<pre><code>query getCommitMessage($repoOwner: String!, $repoName: String!, $sourceBranch: String!, $oid: GitObjectID!) {
  repository(owner: $repoOwner, name: $repoName) {
    ref(qualifiedName: $sourceBranch) {
      repository {
        object(oid: $oid) {
          ... on Commit {
            message
          }
        }
      }
    }
  }
}
</code></pre>
<p>The commit is retrieved regardless which branch I limit it to. I’d expect the commit to only be found if it’s in the given <code>$sourceBranch</code></p>

**Comment by jstacoder** [comment]
<p>i see what happend, you are going back up to the repo, but you are already there.</p>

**Comment by jstacoder** [comment]
<p><a class="mention" href="/sqren">@sqren</a> looks like my original format for the ref was off, because I was going from memory. But yea, it should be <code>refs/heads/BRANCHNAME</code></p>

**Comment by jstacoder** [comment]
<p>sorry i answered without really looking at your question, you are getting the commit regardless because you are asking by oid, which doesnt limit itself by anything, to gaurentee you are in the correct repo you cou ask its name, or get the commit message from the latest commit ie:</p>
<pre><code>query getCommitMessage($repoOwner: String!, $repoName: String!, $sourceBranch: String!) {
  repository(owner: $repoOwner, name: $repoName) {
    ref(qualifiedName: $sourceBranch) {
      name
      target{
        ...on Commit{
          message
        }

      }
    }
  }
}
</code></pre>

**Comment by sorenlouv** [comment]
<p>Hi again,</p>
<blockquote>
<p>to gaurentee you are in the correct repo you cou ask its name, or get the commit message from the latest commit ie:</p>
</blockquote>
<p>Thanks for the suggestion but I need to know whether the commit is on the given branch. Your suggestion seems to only get the head commit.</p>

**Comment by sorenlouv** [comment]
<blockquote>
<p><a class="mention" href="/html-codder">@html-codder</a> wrote:</p>
</blockquote>
<blockquote>
<p>hello Sqren,</p>
</blockquote>
<blockquote>
<p> I hope this message finds someone reading it right, I just want to ask if I can use this commit?</p>
</blockquote>
<p>Not sure which commit you mean but you are welcome to use any of the code snippets in this thread.</p>
